src/Illuminate/Database/Concerns/ManagesTransactions.php

```diff
@@ -34,14 +34,10 @@ public function transaction(Closure $callback, $attempts = 1)
             // If we catch an exception we'll rollback this transaction and try again if we
             // are not out of attempts. If we are out of attempts we will just throw the
             // exception back out and let the developer handle an uncaught exceptions.
-            catch (Exception $e) {
+            catch (Throwable $e) {
```

Comment:
Inside `handleTransactionException` we rollback and re-throw, but we check if transaction is caused by deadlock, this will never be true if the exception is an Error so the behaviour is the same, the only extra part is the check inside `handleTransactionException` to see if the exception is caused by deadlock `causedByDeadlock`